File: src/cpp_pkg/src/publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals; // 启用时间自变量，例如 500ms

class MinimalPublisher : public rclcpp::Node { // 继承 rclcpp::Node 类，表明这是一个 C++ ROS2 节点
  public:
    MinimalPublisher() : Node("minimal_publisher"), count_(0) {
        publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10); // 创建发布者，发布到 topic 话题，队列大小为 10
        timer_ = this->create_wall_timer(
            500ms, std::bind(&MinimalPublisher::callback, this)); // 创建定时器，每 500ms 调用一次 callback 函数
    }

  private:
    void callback() {
        auto message = std_msgs::msg::String();                                    // 创建 std_msgs::msg::String 类型的消息
        message.data = "Hello, World: " + std::to_string(count_++);                // 设置消息内容
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str()); // 打印发布信息
        publisher_->publish(message);                                              // 发布消息
    }

    rclcpp::TimerBase::SharedPtr timer_;                            // 定时器（智能指针管理）
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_; // 发布者（智能指针管理）
    size_t count_;                                                  // 计数器
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);                           // 初始化 ROS2
    rclcpp::spin(std::make_shared<MinimalPublisher>()); // 运行节点，直到节点退出
    rclcpp::shutdown();                                 // 关闭 ROS2
    return 0;
}

File: src/py_pkg/py_pkg/publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class MinimalPublisher(Node):  # 继承 Node 类，表示这是一个 ROS2 节点

    def __init__(self):
        super().__init__("minimal_publisher")  # 初始化节点，命名为 minimal_publisher
        self.publisher_ = self.create_publisher(
            String, "topic", 10
        )  # 创建发布者，发布到 topic 话题，队列大小为 10
        self.timer = self.create_timer(
            0.5, self.callback
        )  # 创建定时器，每隔 0.5 秒执行一次回调函数
        self.count = 1  # 计数器

    def callback(self):
        message = String()  # 创建 String 消息
        message.data = "Hello, World: " + str(self.count)  # 设置消息内容
        self.publisher_.publish(message)  # 发布消息
        self.get_logger().info(f"Publishing: '{message.data}'")  # 打印日志
        self.count += 1


def main(args=None):
    rclpy.init(args=args)  # 初始化 rclpy
    minimal_publisher = MinimalPublisher()  # 创建节点对象
    rclpy.spin(minimal_publisher)  # 运行节点
    minimal_publisher.destroy_node()  # 销毁节点
    rclpy.shutdown()  # 关闭 rclpy


if __name__ == "__main__":
    main()
